### ISSUE TITLE
[discussion] Unite nn modules nn.Something*d to just nn.Something with appropriate options

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#41243]
I proposed it in https://github.com/pytorch/pytorch/issues/41081#issuecomment-655148816, but maybe a separate issue is a better place to discuss this:

A more radical proposal: unite BatchNorm*d / SyncBatchNorm in one module BatchNorm (with option sync = False|True, especially given that SyncBatchNorm is now established in core and is reasonably common). I think now BatchNorm*d are all same except shape checking (and already implemented in a common parent class _BatchNorm). Then switching on sync would become

for bn in [module for module in model.modules() if isinstance(module, torch.nn.BatchNorm)]:
   bn.sync_mode = True # or bn.sync_mode(True)
There may be same proposal to unite Conv*d into Conv, MaxPool*d into MaxPool etc. Dimensionality may be defined by shape of kernel_size.

Fusion code may become simpler, several versions of fused modules (that now separately exist ConvBn*d, LinearBn*d etc) would likely be simplified and unified.

If not for all, maybe for some cases this can be done (like we already have only one nn.SyncBatchNorm, and functional forms are F.batch_norm and not three separate functions)

I think this design already exists in ONNX.

**Comment by fmassa** [comment]
Related to https://github.com/pytorch/pytorch/issues/2628 for BatchNorm

**Comment by mruberry** [comment]
Agreed it's an interesting idea, but not something we're actively looking into.

**Comment by albanD** [comment]
Note that they actually have a shared (internal) subclass (not considering the SyncBatchNorm) which we might be able to leverage here.
But we still need a new common API.

**Comment by vadimkantorov** [comment]
IIRC convolutions also already have some common class like` _ConvNd`. An example of an SqueezeAndExcite1d module that would have benefitted from `arity` polymorphism: https://gist.github.com/vadimkantorov/97d7ff97fd296330a496de5449bd288e

**Comment by vadimkantorov** [comment]
As seen in https://github.com/pytorch/pytorch/issues/41559, it seems that Keras TF interface also does away with just one BatchNormalization class

**Comment by zou3519** [comment]
Related: https://github.com/pytorch/pytorch/issues/6442, we've thought about doing away with all the Dropout{N}d variants and having a single Dropout class before
